src/cache.c

```diff
@@ -0,0 +1,387 @@
+/*
+ * rv32emu is freely redistributable under the MIT License. See the file
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+
+#include <assert.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "cache.h"
+
+#define MIN(a, b) ((a < b) ? a : b)
+#define GOLDEN_RATIO_32 0x61C88647
+#define HASH(val) \
+    (((val) * (GOLDEN_RATIO_32)) >> (32 - (cache_size_bits))) & (cache_size - 1)
+
+static uint32_t cache_size, cache_size_bits;
+
+/*
+ * ARC improves the basic LRU strategy by splitting the cache into two lists,
+ * T1 and T2, the former for recently and the latter for frequently referenced
+ * entries. In turn, each of these is extended with a ghost list (B1 or B2),
+ * which is attached to the bottom of the two lists. When a cache hit occurs
+ * in B1, it indicates that T1's capacity is too little, therefore we increase
+ * T1's size while decreasing T2. But, if the cache hit occurs in B2, we would
+ * increase the size of T2 and decrease the size of T1.
+ */
+typedef enum {
+    LRU_list,
+    LFU_list,
+    LRU_ghost_list,
+    LFU_ghost_list,
+    N_CACHE_LIST_TYPES
+} cache_list_t;
+
+struct list_head {
+    struct list_head *prev, *next;
+};
+
+struct hlist_head {
+    struct hlist_node *first;
+};
+
+struct hlist_node {
+    struct hlist_node *next, **pprev;
+};
+
+typedef struct {
+    void *value;
+    uint32_t key;
+    cache_list_t arc_type;
+    struct list_head list;
```

Comment:
Is it feasible to drop the use of `list` here?

Reply:
We cannot manipulate the doubly-linked list withot `list`.

Reply:
`list` is for four cache lists T1, T2, B1, and B2 and `ht_list` is for hash table, we should find out the certain cache entry by hash table.

Reply:
Add corresponding comments along with further improvements.